### DISCUSSION TITLE
Allocated memory increases over time (possible memory leak?)

### DISCUSSION THREAD

**Discussion by devandreassimmeth** [fastapi/fastapi#11079] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
---
```

### Description

Hello,

In our company, we are using multiple FastAPI applications as backend deployed in a Docker-environment on an AWS EC2 instance. As database we use a MongoDB with Beanie/Motor for the database calls. Because our collections use schema-versioning, we need a mapping before the data can be loaded into the corresponding data-class and that's why we often use Motor calls instead of the direct Beanie calls.

So far, so good. Some weeks ago, we discovered that the allocated working memory in our backends is increasing day by day. So we went in search of where this was coming from and our first thought was, that it comes from the "AsyncIOMotorCollection" of "Motor" library. Because of that we created a reproduceable example and created an issue with the example in the Motor Jira-Board: https://jira.mongodb.org/browse/MOTOR-1231

In this issue, you also find a much deeper description of the memory allocation, the example application and the findings made so far. The MongoDB team also already investigated time and found out, that it's not database related, because the allocation is also there when using a PostgreSQL database. Their strong assumption was then, that it is because of something in FastAPI and have recommended that I open an issue here for this. Since this is unfortunately not possible, I am trying it here. Maybe someone has an explanation for the behaviour we are observing.

Thanks a lot for the help!

All the best,
Andreas

### Operating System

Windows

### Operating System Details

Also when running on Ubuntu 22.04 Server in Docker.

### FastAPI Version

0.108.0

### Pydantic Version

2.6.0

### Python Version

3.10.8

**Comment by n8sty** [comment]
There have been some long running discussions/issues about this in the past, see https://github.com/tiangolo/fastapi/discussions/8612 for possibly the most extensive. However, I don't think you'll get much help if you a reproducible example isn't provided. I know for these kinds of issues that's tough.

**Reply by devandreassimmeth** [reply]
Hey @n8sty,

Thanks for your answer and the link. I already saw that thread, yes. And the behaviour in the chart provided in this thread looks the same as we and the guys from the "Motor"-library could reproduce.

And I don't know how good you read over my text, but I provided a reproducible example. But not directly in the section here, because it's a multi-file example using poetry for dependency management hanging in the "Motor"-issue linked in my text. And both, I and the "Motor"-team, could reproduce the behaviour with python 3.10 AND the newest FastAPI version in this small example application. So it's definitely a current problem we are facing.

I know the toughness of these kind of issues, of course, and it's even tougher when they come from code, you didn't implement by yourself. And that's why I'm searching for help here, from the experts who are deep in the code of the library where it most likely comes from according to current knowledge.

Andreas :)

**Reply by ryanrib14** [reply]
Some update about this @devandreassimmeth ? i also have the same problem here ...

**Comment by rumbarum** [comment]
@devandreassimmeth 
Any news about this? 
I have a same issue with postgresql. I also have same problem. 
Any hint would be appreciated!

**Comment by thinkabout4x** [comment]
Hello! About memory leak. I faced the same issue and found the problem to be in uvicorn programmatic run.

I have main.py file which spawn uvicorn and ./app/app.py contain fastapi app definition.

In the main.py before i had:
```
from app.app import app

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 4000))
    env = os.environ.get("ENV", "unspecified")
    host = os.environ.get("HOST", "0.0.0.0")
    if env.casefold() == "prod".casefold():
        uvicorn.run("main:app", host=host, port=port, workers=4)
    elif env.casefold() == "local".casefold():
        uvicorn.run("main:app", host=host, port=port, reload=True)
    else:
        uvicorn.run("main:app", host=host, port=port, workers=4)
```
This setup is generating memory leak and I assume because of improper access of the app (it is not directly used anywhere) and uvicorn string "main:app" invokes second instance of app which leads to problem.

Now fixed setup that works without memory leak in my case:
```
if __name__ == "__main__":
    port = int(os.environ.get("PORT", 4000))
    env = os.environ.get("ENV", "unspecified")
    host = os.environ.get("HOST", "0.0.0.0")
    if env.casefold() == "prod".casefold():
        uvicorn.run("app.app:app", host=host, port=port, workers=4)
    elif env.casefold() == "local".casefold():
        uvicorn.run("app.app:app", host=host, port=port, reload=True)
    else:
        uvicorn.run("app.app:app", host=host, port=port)
```
No import here, uvicorn string correctly points to app location "app.app:app"

Versions: fastapi = 0.115.12, uvicorn = 0.21.1

**Comment by nfergu** [comment]
As mentioned [here](https://github.com/pydantic/pydantic/issues/9429#issuecomment-3458905343), I don't think the issue in [MOTOR-1231](https://jira.mongodb.org/browse/MOTOR-1231) is related to FastAPI.

Here's a partial graph of the referrers of one of the leaked objects (produced by [memalot](https://pypi.org/project/memalot/)):

```
╙── ExampleObject (object) (id=16737063056) 
    └── list index=4999 (id=13496758976) 
        └── list (object) (id=13496758976) 
            └── Example.obj_array (instance attribute) (id=16737063136) 
                └── Example (object) (id=16737063136) 
                    └── CreateExampleRequest.example (instance attribute) (id=15184393712) 
                        └── CreateExampleRequest (object) (id=15184393712) 
                            └── tuple index=1 (id=13357105344) 
                                └── tuple (object) (id=13357105344) 
                                    └── dict (object) (id=4459673920) 
                                        └── _lru_cache_wrapper (object) (id=4458998672) 
                                            └── dict key=_resolve_pipeline_behaviors (id=4459673216) 
``` 

The `_resolve_pipeline_behaviors` function is in mediatpy, which has an [unbounded cache](https://github.com/panicoenlaxbox/mediatpy/blob/4c0aec990717569aa7d877b64f733a6b171cac83/src/mediatpy/__init__.py#L271) that (I think) caches every request
